### PR TITLE
[BPK-2326] Implement font theming for label

### PR DESCRIPTION
**Pull request by georgegillams**
+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: Backpack/Font/Classes/Generated/BPKFont.m [hunk 1]
```diff
@@ -187,7 +189,7 @@ + (NSAttributedString *)attributedStringWithFontStyle:(BPKFontStyle)fontStyle co
 
 + (NSDictionary<NSAttributedStringKey, id> *)attributesForFontStyle:(BPKFontStyle)fontStyle
```
Comment (georgegillams) [inline Backpack/Font/Classes/Generated/BPKFont.m]: TODO add `fontMapping` to this method

File: Backpack/Font/Classes/Generated/BPKFont.m [hunk 2]
```diff
@@ -187,7 +189,7 @@ + (NSAttributedString *)attributedStringWithFontStyle:(BPKFontStyle)fontStyle co
...
                                                withCustomAttributes:(NSDictionary<NSAttributedStringKey,id> *)customAttributes {
-    NSMutableDictionary<NSAttributedStringKey, id> *attributes = [[self attributesForFontStyle:fontStyle] mutableCopy];
+    NSMutableDictionary<NSAttributedStringKey, id> *attributes = [[self attributesForFontStyle:fontStyle fontMapping:nil] mutableCopy];
```
Comment (georgegillams) [inline Backpack/Font/Classes/Generated/BPKFont.m]: TODO pass `fontMapping` down

File: templates/BPKFont.m.njk [hunk 3]
```diff
@@ -140,7 +146,31 @@ NS_ASSUME_NONNULL_BEGIN
...
+             {% else -%}
+               return [UIFont fontWithName:fontMapping.regular size:{{f.size}}];
+             {% endif %}{% endfor %}
+            default:
+              NSAssert(NO, @"Unknown fontStyle %ld", (unsigned long)style);
+    }
+}
+
++ (NSNumber *_Nullable)trackingForStyle:(BPKFontStyle)style andFont:(NSString *)font {
+    if (font != nil) {
```
Comment (georgegillams) [inline templates/BPKFont.m.njk]: This ain't right yo!

File: Backpack/Font/Classes/BPKFontMapping.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ * @param bold the font family for bold text
+ * @param regular the font family for regular text
+ * @param heavy the font family for heavy text
+ * @return `BPKFontMapping` instance.
+ */
+- (instancetype)initWithFamily:(NSString *)regular
+                withBoldFamily:(NSString *)bold
+               withHeavyFamily:(NSString *)heavy NS_DESIGNATED_INITIALIZER;
+
+@property(nonatomic, strong) NSString *regular;
```
Comment (k0nserv) [inline Backpack/Font/Classes/BPKFontMapping.h]: Why are these `readwrite`? Shouldn't they be `readonly`?

Comment (k0nserv) [reply Backpack/Font/Classes/BPKFontMapping.h]: Can we consider more descriptive names? Something `regularFontFace`?

File: Backpack/Font/Classes/BPKFontMapping.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * Creates a `BPKSwitch` with a decoder (typically when creating from storyboards)
+ *
+ * @param bold the font family for bold text
+ * @param regular the font family for regular text
+ * @param heavy the font family for heavy text
+ * @return `BPKFontMapping` instance.
+ */
+- (instancetype)initWithFamily:(NSString *)regular
```
Comment (k0nserv) [inline Backpack/Font/Classes/BPKFontMapping.h]: A family of fonts is a collection of font faces and presumably all of these will come from the same family so maybe we should use the term `fontFace` instead?

File: Backpack/Font/Classes/BPKFontMapping.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+@property(nonatomic, strong) NSString *regular;
+@property(nonatomic, strong) NSString *semibold;
+@property(nonatomic, strong) NSString *heavy;
+@property(nonatomic, strong) NSString *base;
+
+/**
+ * The font family to use for a given weight.
+ */
+- (NSString *)familyForWeight:(NSString *)weight;
```
Comment (k0nserv) [inline Backpack/Font/Classes/BPKFontMapping.h]: Why is `weight` a `NSString`?

File: Example/Tests/BPKFontTest.m [hunk 7]
```diff
@@ -56,7 +56,8 @@ - (void)testAttributesForFontStyleWithCustomAttributes {
                                               NSKernAttributeName: @1.5,
                                               NSFontAttributeName: font,
                                               NSForegroundColorAttributeName: UIColor.redColor,
-                                          }];
+                                          }
```
Comment (k0nserv) [inline Example/Tests/BPKFontTest.m]: We need more tests here. For example when a `fontMapping`  is provided

File: Backpack/Font/Classes/BPKFontMapping.m [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        _family = [family copy];
+        _regularFontFace = [regularFontFace copy];
+        _semiboldFontFace = [semiboldFontFace copy];
+        _heavyFontFace = [heavyFontFace copy];
+    }
+    return self;
+}
+
+- (NSString *)familyForWeight:(UIFontWeight)weight {
+    if (weight == UIFontWeightRegular) {
```
Comment (k0nserv) [inline Backpack/Font/Classes/BPKFontMapping.m]: Keep in mind that `UIFontWeight` is a actually a type aliased `CGFloat` so you are comparing floats here. Not so sure about that

File: Backpack/Font/Classes/Generated/BPKFont.m [hunk 9]
```diff
@@ -17,17 +17,19 @@
...
 NS_ASSUME_NONNULL_BEGIN
 
 @interface BPKFont()
 @property(nonatomic, strong, readonly) NSCache<NSString *, NSDictionary *> *attributesCache;
 
-+ (NSString *)cacheKeyForFontStyle:(BPKFontStyle)style;
++ (NSString *)cacheKeyForFontStyle:(BPKFontStyle)style andFontMapping:(BPKFontMapping *_Nullable)fontMapping;
 + (UIFont *)fontForStyle:(BPKFontStyle)style;
-+ (NSNumber *_Nullable)trackingForStyle:(BPKFontStyle)style;
++ (NSNumber *_Nullable)trackingForStyle:(BPKFontStyle)style fontMapping:(BPKFontMapping *_Nullable)fontMapping;
```
Comment (k0nserv) [inline Backpack/Font/Classes/Generated/BPKFont.m]: We can remove private methods, they don't need to be declared

File: Backpack/Font/Classes/Generated/BPKFont.m [hunk 10]
```diff
@@ -156,15 +158,15 @@ + (UIFont *)textXxxlHeavy {
 }
 
 
-+ (NSAttributedString *)attributedStringWithFontStyle:(BPKFontStyle)fontStyle content:(NSString *)content {
-    NSDictionary *attributes = [self attributesForFontStyle:fontStyle];
++ (NSAttributedString *)attributedStringWithFontStyle:(BPKFontStyle)fontStyle content:(NSString *)content fontMapping:(BPKFontMapping *_Nullable)fontMapping {
```
Comment (k0nserv) [inline Backpack/Font/Classes/Generated/BPKFont.m]: Maybe split this declaration on multiple lines?

File: Backpack/Font/Classes/Generated/BPKFont.m [hunk 11]
```diff
@@ -156,15 +158,15 @@ + (UIFont *)textXxxlHeavy {
...
+    NSDictionary *attributes = [self attributesForFontStyle:fontStyle fontMapping:fontMapping];
     NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:content attributes:attributes];
 
     return attributedString;
 }
 
-+ (NSAttributedString *)attributedStringWithFontStyle:(BPKFontStyle)fontStyle content:(NSString *)content textColor:(UIColor *)textColor {
-    NSMutableDictionary *attributes = [[self attributesForFontStyle:fontStyle] mutableCopy];
++ (NSAttributedString *)attributedStringWithFontStyle:(BPKFontStyle)fontStyle content:(NSString *)content textColor:(UIColor *)textColor fontMapping:(BPKFontMapping *_Nullable)fontMapping {
```
Comment (k0nserv) [inline Backpack/Font/Classes/Generated/BPKFont.m]: Maybe split this declaration on multiple lines?

File: Backpack/Font/Classes/BPKFontMapping.m (hunk 8 above)
Comment (georgegillams) [reply Backpack/Font/Classes/BPKFontMapping.m]: @k0nserv all done now. Not sure if I changed this the way you expected? https://github.com/Skyscanner/backpack-ios/pull/306/files#diff-f0207de3091521efb61f277e6b8ec8f3R42

**Pull request merged**